### PR TITLE
chore(dao): Add explicit ON DELETE CASCADE when deleting datasets

### PR DESCRIPTION
**Pull request by john-bodley**
### SUMMARY

Per  [SIP-92 Proposal for restructuring the Python code base](https://github.com/apache/superset/issues/20630) https://github.com/apache/superset/pull/24331 organized all the DAOs to be housed within a shared folder—the result of which highlighted numerous inconsistencies, repetition, and inefficiencies.

Personally I really struggle at times with the SQLAlchemy model especially with regards to the relationship logic and why the cascade paradigm only impacts the SQLAlchemy ORM, i.e., isn't materialized to the actual table schema. The TL;DR is there always seems to be multiple ways to skin the cat which just adds to the confusion.

Something clearly is quite right with our model as it's often the case when deleting a dataset we first delete the owners, columns, and metrics to avoid foreign key constrain violations. I gather this is because of our bulk delete logic (see [this](https://apache-superset.slack.com/archives/G013HAE6Y0K/p1687396791699689) discussion for more detail) bypasses the SQLAlchemy ORM or there are other gremlins at play—there is [this](https://stackoverflow.com/questions/72223166/sqlalchemy-delete-rows-from-associate-table-from-many-to-many-relationship) issue reported on Stack Overflow regarding cascading for association tables.

I recently added the [DAO-Style-Guidelines-and-Best-Practices](https://github.com/apache/superset/wiki/DAO-Style-Guidelines-and-Best-Practices) wiki page, and I hope there is general agreement in item 5 which mentions the "shift left" mentality where the data model should persist in the database schema than rely on Python logic per the SQLAlchemy ORM. The beauty of this approach is it simplifies the codebase and mental model and ensures that the correct logic is adhered to if/when executing SQL outside of the ORM (be that using the SQLAlchemy bulk workflows, executing SQL against the DB-API, using the MySQL/PostgreSQL CLI, etc.).

The logic I employed follows [this](https://docs.sqlalchemy.org/en/20/orm/cascades.html#using-foreign-key-on-delete-cascade-with-orm-relationships) SQLAlchemy documentation. Per their suggestion I opted for `passive_deletes=True` as we don't require the ORM to load the columns, metrics, and owners (if lazy loaded) when deleting the dataset.

The TL;DR is prior to adding the migration CI was failing so it seems like this approach feels somewhat correct.

### TESTING INSTRUCTIONS

CI.

### ADDITIONAL INFORMATION

- [ ] Has associated issue: https://github.com/apache/superset/issues/20630
- [ ] Required feature flags:
- [ ] Changes UI
- [x] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [x] Migration is atomic, supports rollback & is backwards-compatible
  - [x] Confirm DB migration upgrade and downgrade tested
  - [x] Runtime estimates and downtime expectations provided: Takes ~ 1 minute for Airbnb's instance w/o load.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/daos/dataset.py [hunk 1]
```diff
@@ -364,17 +364,6 @@ def create_metric(
     @staticmethod
     def bulk_delete(models: Optional[list[SqlaTable]], commit: bool = True) -> None:
         item_ids = [model.id for model in models] if models else []
-        # bulk delete, first delete related data
```
Comment (john-bodley) [inline superset/daos/dataset.py:367]: I'm not sure if this was required because when you bulk delete a table (using the `delete()` method) you bypass the ORM and that's where the cascade logic resides, but given we've adopted the "shift left" mentality it's something of a moot point, i.e., the database will handle the deletion irrespective of how the records are deleted (SQLAlchemy ORM, raw connection, MySQL/PostgreSQL CLI, etc.).

File: superset/datasets/commands/delete.py [hunk 2]
```diff
@@ -43,19 +41,13 @@ def __init__(self, model_id: int):
 
     def run(self) -> Model:
         self.validate()
-        self._model = cast(SqlaTable, self._model)
+        assert self._model
+
         try:
-            # Even though SQLAlchemy should in theory delete rows from the association
```
Comment (john-bodley) [inline superset/datasets/commands/delete.py:48]: @betodealmeida this should address the problem you were trying to circumvent in https://github.com/apache/superset/pull/23414.

Fun fact I wondered if I wrote this comment as it's somewhat in the style I write, but I realized I don't typically use the term "let's".

File: tests/integration_tests/sqla_models_tests.py [hunk 3]
```diff
@@ -189,11 +189,6 @@ def test_extra_cache_keys(self, flask_g):
         self.assertTrue(table3.has_extra_cache_key_calls(query_obj))
         assert extra_cache_keys == ["abc"]
 
-        # Cleanup
```
Comment (john-bodley) [inline tests/integration_tests/sqla_models_tests.py:192]: CI was failing with a SQLAlchemy error which would indicate that we're trying to delete objects which previously weren't committed to the database. Given how these tables are constructed, 

```
table3 = SqlaTable(
            table_name="test_has_no_extra_cache_keys_table",
            sql=query,
            database=get_example_database(),
        )
```

and there's no clear add/merge call it seems that SQLAlchemy's viewpoint is right, i.e., there's no cleanup required. If I'm write then this is a win for SIP-92. I suspect throughout the code and tests we're doing numerous database operations inefficiently with superfluous commits, not rolling back test state, etc.

File: superset/datasets/commands/bulk_delete.py [hunk 4]
```diff
@@ -40,35 +39,10 @@ def __init__(self, model_ids: list[int]):
 
     def run(self) -> None:
         self.validate()
-        if not self._models:
-            return None
+        assert self._models
+
         try:
             DatasetDAO.bulk_delete(self._models)
-            for model in self._models:
```
Comment (john-bodley) [inline superset/datasets/commands/bulk_delete.py:47]: This is logic taken https://github.com/apache/superset/pull/24466. Previously a test was failing because the relationship no longer existed in Python after the commit. See [here](https://docs.sqlalchemy.org/en/20/orm/cascades.html#notes-on-delete-deleting-objects-referenced-from-collections-and-scalar-relationships) for details. The issue was that:

```python
DatasetDAO.bulk_delete(self._models)
```

includes a `db.session.commit()` in which all the attributes have expired. I'm not really sure why this worked before, but the TL;DR is this logic has been moved inside the DAO and is executed prior to the commit.

File: superset/utils/core.py [hunk 5]
```diff
@@ -815,12 +815,15 @@ def __exit__(  # pylint: disable=redefined-outer-name,redefined-builtin
...
 def pessimistic_connection_handling(some_engine: Engine) -> None:
     @event.listens_for(some_engine, "engine_connect")
-    def ping_connection(connection: Connection, branch: bool) -> None:
+    def on_connect(connection: Connection, branch: bool) -> None:
         if branch:
             # 'branch' refers to a sub-connection of a connection,
             # we don't want to bother pinging on these.
             return
 
+        if connection.dialect.name == "sqlite":
```
Comment (john-bodley) [inline superset/utils/core.py:824]: See [here](https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#foreign-key-support) for why this is needed. I do wonder if we should drop support for SQLite given it's not supported in production. We already skip a bunch of tests, i.e., 

```python
if backend() == "sqlite":
    return
```

possibly because of how the foreign key constraints have no impact on the underlying table.

File: superset/utils/core.py [hunk 6]
```diff
@@ -849,6 +851,19 @@ def ping_connection(connection: Connection, branch: bool) -> None:
             # restore 'close with result'
             connection.should_close_with_result = save_should_close_with_result
 
+        if some_engine.dialect.name == "sqlite":
```
Comment (john-bodley) [inline superset/utils/core.py:854]: I wonder if this is why a number of tests currently have,

```python
if backend() == "sqlite":
    return
```

File: tests/integration_tests/sqla_models_tests.py [hunk 7]
```diff
@@ -430,7 +425,7 @@ def test_multiple_sql_statements_raises_exception(self):
         }
 
         table = SqlaTable(
-            table_name="test_has_extra_cache_keys_table",
+            table_name="test_multiple_sql_statements",
```
Comment (john-bodley) [inline tests/integration_tests/sqla_models_tests.py:428]: See my comment above. SQLite was complaining that said table already existed when I removed the "cleanup", though neither MySQL nor PostgreSQL did. I'm not entirely sure what's going on here, but it does seem like the table name was copypasta from the other test and thus the easiest solution was to rename the table in accordance with the test. 

I sense with this refactoring work it's a case of two steps forwards, one step back at time.

File: superset/utils/core.py [hunk 8]
```diff
@@ -849,6 +850,24 @@ def ping_connection(connection: Connection, branch: bool) -> None:
...
+
+            @event.listens_for(some_engine, "connect")
+            def set_sqlite_pragma(  # pylint: disable=unused-argument
```
Comment (john-bodley) [inline superset/utils/core.py:856]: Hopefully (if approved) this will remove the need for [this](https://github.com/search?q=repo%3Aapache%2Fsuperset+backend%28%29+%3D%3D+%22sqlite%22&type=code) logic in a number of tests.

**Comment by craig-rueda** [comment]
Totally agree with this philosophy, except for situations where you really want to leave FK data hanging. IMO, the DB should keep things clean wherever possible and should enforce referential integrity (kinda the point of FK's in the first place :D ). Delegating to the DB to deal with cleanup keeps app-tier logic cleaner and less buggy as it's less concerned with cascades and what not.

File: superset/daos/dataset.py [hunk 9]
```diff
@@ -361,25 +362,24 @@ def create_metric(
...
             db.session.query(SqlaTable).filter(SqlaTable.id.in_(item_ids)).delete(
                 synchronize_session="fetch"
             )
+
+            if models:
+                connection = db.session.connection()
+                mapper = next(iter(cls.model_cls.registry.mappers))  # type: ignore
+
+                for model in models:
+                    security_manager.dataset_after_delete(mapper, connection, model)
```
Comment (craig-rueda) [inline superset/daos/dataset.py:381]: Ideally, we wouldn't reference the SM from the DAO layer as it's more of a mid-tier concern. This belongs in the CMD.

File: superset/migrations/versions/2023-06-22_13-39_6fbe660cac39_add_on_delete_cascade_for_tables_references.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    """
+
+    bind = op.get_bind()
+    insp = sa.engine.reflection.Inspector.from_engine(bind)
+
+    conv = {
+        "fk": "fk_%(table_name)s_%(column_0_name)s_%(referred_table_name)s",
+    }
+
+    for table in ("sql_metrics", "table_columns"):
```
Comment (craig-rueda) [inline superset/migrations/versions/2023-06-22_13-39_6fbe660cac39_add_on_delete_cascade_for_tables_references.py:52]: 👍

File: superset/daos/dataset.py (hunk 9 above)
Comment (john-bodley) [reply superset/daos/dataset.py:381]: @craig-rueda could you help educate me as to why this is? Granted that the DAO is only abstraction from the database (and not the security manager which in turns interfaces with the database), but isn't this the interface we expect commands, APIs, etc. to interface with? Note this logic would be handled by the `after_delete` event listener if we were using the ORM, as opposed to bulk queries, to delete the assets.

Comment (john-bodley) [reply superset/daos/dataset.py:381]: @craig-rueda I generally agree, but how does this differ from the DAO asking "please delete these records" which triggers the SQLAlchemy ORM to handle post deletion events (per the [after_delete](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.MapperEvents.after_delete) event handler) which in turn invokes the security manager?

Note we typically use these SQLAlchemy events because of short comings with either our data model and/or constructs which can't be modeled by our database schema.

Comment (craig-rueda) [reply superset/daos/dataset.py:381]: Yea, either way. It's more of a stylistic thing IMO. Keeping the DAO layer as decoupled as possible has worked well for me in the past, hence my comments :).

**Review by craig-rueda — APPROVED** [review]
